File: FanPulse_AOS/data/src/main/java/com/aos/fanpulse/data/remote/dto/Chart.kt
package com.aos.fanpulse.data.remote.dto

// 차트 전체 정보를 담는 모델
data class ChartDetail(
    val id: String,                 // 차트 ID (UUID)
    val chartType: String,          // 차트 종류 (예: DAILY, WEEKLY, REALTIME)
    val chartDate: String,          // 차트 기준 날짜 ("2026-04-01")
    val entries: List<ChartEntry>,  // 차트에 포함된 순위 리스트
    val createdAt: String           // 데이터 생성 일시
)

// 차트 내 개별 순위 항목
data class ChartEntry(
    val id: String,                 // 엔트리 고유 ID
    val rank: Int,                  // 현재 순위
    val trackId: String?,           // 곡 ID (곡 정보가 없는 차트일 경우 null 가능)
    val artistId: String,           // 아티스트 ID
    val trackTitle: String?,        // 곡 제목
    val artistName: String,         // 아티스트 이름
    val previousRank: Int,          // 이전 순위
    val peakRank: Int,              // 최고 순위
    val weeksOnChart: Int,          // 차트 진입 주차
    val rankChange: Int,            // 순위 변동폭 (예: +2, -1)
    val isNew: Boolean              // 신규 진입 여부
)

enum class ChartType {
    MELON, BUGS, GENIE, FLO, VIBE,
    BILLBOARD_KR, BILLBOARD_US,
    SPOTIFY, APPLE_MUSIC
}

data class ChartHistoryResponse(
    val content: List<ChartHistoryItem>
)

data class ChartHistoryItem(
    val id: String,                 // 특정 날짜의 차트 ID (상세 조회 시 사용)
    val chartType: String,          // 차트 플랫폼 (MELON 등)
    val chartDate: String,          // 차트 날짜 ("2026-04-01")
    val entryCount: Int             // 해당 차트에 등록된 순위 데이터 개수
)

File: web/src/lib/api/chart.ts
import { apiClient } from '@/lib/api-client';
import {
  isIsoDate,
  isIsoInstant,
  isNonEmptyString,
  isNullableInteger,
  isPositiveInteger,
  isRecord,
  isUuid,
  unwrapApiResponse,
} from '@/lib/api-response';
import type { ApiResponse } from '@/types/api';

export type ChartType =
  | 'MELON'
  | 'BUGS'
  | 'GENIE'
  | 'FLO'
  | 'VIBE'
  | 'BILLBOARD_KR'
  | 'BILLBOARD_US'
  | 'SPOTIFY'
  | 'APPLE_MUSIC';

const CHART_TYPES = new Set<ChartType>([
  'MELON',
  'BUGS',
  'GENIE',
  'FLO',
  'VIBE',
  'BILLBOARD_KR',
  'BILLBOARD_US',
  'SPOTIFY',
  'APPLE_MUSIC',
]);

export interface ChartEntry {
  id: string;
  rank: number;
  trackId: string;
  artistId: string | null;
  trackTitle: string;
  artistName: string;
  previousRank: number | null;
  peakRank: number;
  weeksOnChart: number;
  rankChange: number | null;
  isNew: boolean;
  artworkUrl: string | null;
}

export interface ChartResponse {
  id: string;
  chartType: ChartType;
  chartDate: string;
  entries: ChartEntry[];
  createdAt: string;
}

function isNullableAppleArtworkUrl(value: unknown): value is string | null {
  if (value === null) return true;
  if (typeof value !== 'string') return false;
  try {
    const url = new URL(value);
    const pathname = url.pathname;
    const rawPath = url.href.slice(url.origin.length).split('?')[0].split('#')[0];
    return (
      url.protocol === 'https:' &&
      /^is[1-9]-ssl\.mzstatic\.com$/.test(url.hostname) &&
      (url.port === '' || url.port === '443') &&
      url.username === '' &&
      url.password === '' &&
      url.search === '' &&
      url.hash === '' &&
      pathname.startsWith('/image/thumb/') &&
      !pathname.includes('/../') &&
      !pathname.includes('/./') &&
      !/%2e|%2f|%5c/i.test(rawPath) &&
      (pathname.endsWith('.jpg') || pathname.endsWith('.jpeg') || pathname.endsWith('.png'))
    );
  } catch {
    return false;
  }
}

function isChartEntry(value: unknown): value is ChartEntry {
  if (!isRecord(value)) return false;
  return (
    isUuid(value.id) &&
    isPositiveInteger(value.rank) &&
    isUuid(value.trackId) &&
    (value.artistId === null || isUuid(value.artistId)) &&
    isNonEmptyString(value.trackTitle) &&
    isNonEmptyString(value.artistName) &&
    (value.previousRank === null || isPositiveInteger(value.previousRank)) &&
    isPositiveInteger(value.peakRank) &&
    isPositiveInteger(value.weeksOnChart) &&
    isNullableInteger(value.rankChange) &&
    typeof value.isNew === 'boolean' &&
    isNullableAppleArtworkUrl(value.artworkUrl)
  );
}

function isChartResponse(value: unknown): value is ChartResponse {
  if (!isRecord(value)) return false;
  return (
    isUuid(value.id) &&
    typeof value.chartType === 'string' &&
    CHART_TYPES.has(value.chartType as ChartType) &&
    isIsoDate(value.chartDate) &&
    Array.isArray(value.entries) &&
    value.entries.every(isChartEntry) &&
    isIsoInstant(value.createdAt)
  );
}

export async function fetchLatestChart(
  chartType: ChartType,
  signal?: AbortSignal
): Promise<ChartResponse> {
  const response = await apiClient.get<ApiResponse<ChartResponse>>(
    `/charts/${chartType}/latest`,
    { signal }
  );

  return unwrapApiResponse(
    response.data,
    '차트 API 응답이 올바르지 않습니다.',
    isChartResponse
  );
}
